pennylane/ops/op_math/composite.py

```diff
@@ -131,14 +134,67 @@ def is_hermitian(self):
     def has_matrix(self):
         return all(op.has_matrix for op in self)
 
-    @abc.abstractmethod
     def eigvals(self):
-        """Return the eigenvalues of the specified operator."""
+        """Return the eigenvalues of the specified operator.
+
+        This method uses pre-stored eigenvalues for standard observables where
+        possible and stores the corresponding eigenvectors from the eigendecomposition.
+
+        Returns:
+            array: array containing the eigenvalues of the operator
+        """
+        eigvals = []
+        for ops in self.overlapping_ops:
+            if len(ops) == 1:
+                eigvals.append(
+                    qml.utils.expand_vector(ops[0].eigvals(), list(ops[0].wires), list(self.wires))
+                )
+            else:
+                tmp_composite = self.__class__(*ops)
+                eigvals.append(
+                    qml.utils.expand_vector(
+                        tmp_composite.eigendecomposition["eigval"],
+                        list(tmp_composite.wires),
+                        list(self.wires),
+                    )
+                )
+
+        return self._math_op(eigvals, axis=0)
 
     @abc.abstractmethod
     def matrix(self, wire_order=None):
         """Representation of the operator as a matrix in the computational basis."""
 
+    @property
+    def overlapping_ops(self) -> List[List[Operator]]:
+        """Groups all operands of the composite operator that act on overlapping wires taking
+        into account operator commutivity.
+
+        Returns:
+            List[List[Operator]]: List of lists of operators that act on overlapping wires. All the
+            inner lists commute with each other.
+        """
+        if self._overlapping_ops is None:
+            overlapping_ops = []  # [(wires, [ops])]
+            for op in self:
+                op_idx = False
+                ops = [op]
+                wires = op.wires
+                for idx, (wire_keys, _) in reversed(list(enumerate(overlapping_ops))):
+                    if any(wire in wire_keys for wire in wires):
+                        old_wires, old_ops = overlapping_ops.pop(idx)
+                        ops = old_ops + ops
+                        wires = old_wires + wires
+                        op_idx = idx
+                if op_idx is not False:
+                    overlapping_ops.insert(op_idx, (wires, ops))
+                else:
+                    overlapping_ops += [(wires, ops)]
+
+            self._overlapping_ops = [overlapping_op[1] for overlapping_op in overlapping_ops]
+
+        return self._overlapping_ops
```

Comment:
We only care about operator commutativity when working with Products right? For the Sum class I imagine we can group arbitrarily. In general, I imagine that this function will depend on the restrictions of the composite operator that is being implemented. 

Maybe we need to make this a little more general? and then allow for composite operators to overwrite this method and include their restrictions (like preserving commutativity)

Reply:
Done!